### PR TITLE
DAOS-6768 bio: Use BIO DMA buffer for SCM RDMA

### PR DESCRIPTION
**Pull request by NiuYawei**
Since SCM MR is disallowed in recent kernel, this patch changed
the default "direct SCM RDMA" mode to "RDMA to DMA buffer + a
SCM <-> buffer memory copy" mode. The "direct SCM RDMA" mode can
be enabled by set the "DAOS_SCM_RDMA_ENABLED=1".

Following changes are made for this new SCM RDMA mode:

- Setup DMA buffer even if NVMe isn't configured;
- Adjusted dma_rw() to support SCM. To simplify NVMe I/O code, the
  "Zeroing -> Modify -> Write" on partial page update is removed;
- Default SCM RDMA won't bypass bulk cache now;
- Dedup memcmp mode now allocates DRAM buffer for data verify, then
  copy data to newly allocated SCM if the verify failed;

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### CODE REVIEW

**Comment by NiuYawei** [comment]
Test failures are degraded EC failures from DAOS-7655. Needs to merge master once DAOS-7655 fixed.

**Comment by NiuYawei** [comment]
Few degraded EC failures in last round of CI, merge master to pick fix from DAOS-7350

**Comment by NiuYawei** [comment]
Only single EC aggregation failure from DAOS-7204.

**Comment by NiuYawei** [comment]
Ping reviewers.

File: src/bio/bio_buffer.c [hunk 1]
```diff
@@ -605,7 +642,8 @@ dma_map_one(struct bio_desc *biod, struct bio_iov *biov, void *arg)
 	last_rg = iod_last_region(biod);
 
 	/* First, try consecutive reserve from the last reserved region */
-	if (last_rg) {
+	if (last_rg && bio_iov2media(biov) != DAOS_MEDIA_SCM &&
+	    bio_iov2media(biov) == last_rg->brr_media) {
```
Comment (liuxuezhao) [inline src/bio/bio_buffer.c:646]: just curious that is it possible bio_iov2media(biov) != last_rg->brr_media?

**Review by liuxuezhao — APPROVED** [review]

File: src/bio/bio_buffer.c (hunk 1 above)
Comment (NiuYawei) [reply src/bio/bio_buffer.c:646]: Yes, that's quite possible. For example, a fetch request to fetch [0, 4k), on VOS storage, the [0, 4k) could consist of [0, 1k)@epoch 2 on SCM and [1k, 4k)@epoch 1 on NVMe.

**Comment by johannlombardi** [comment]
please inspect.

**Review by wangdi1 — APPROVED** [review]

**Pull request merged**